### PR TITLE
[WIP] Fix handling of inline and preformatted tags

### LINKED ISSUE #20: Handling long text and inline elements
Currently, the plugin breaks at the start and end of child content when it can't fit on one line:

```diff
--- input.html	2019-05-12 16:49:09.000000000 -0600
+++ output.html	2019-05-12 16:49:09.000000000 -0600
@@ -1 +1,5 @@
-<span>Lorem ipsum dolor sit amet consectetur adipisicing elit. Tenetur ad fugit cupiditate vel ullam voluptatum maxime harum nemo ipsa sequi sapiente assumenda vitae, beatae iste sed veritatis ipsam dicta deserunt.</span>
+<span>
+    Lorem ipsum dolor sit amet consectetur adipisicing elit. Tenetur ad fugit
+    cupiditate vel ullam voluptatum maxime harum nemo ipsa sequi sapiente
+    assumenda vitae, beatae iste sed veritatis ipsam dicta deserunt.
+</span>
```

Usually, this isn't a problem, but it can be an issue if the parent element (in the above example, `<span>`) has a `display` CSS property with a value of `inline`. In that case, an extra space will be present between sibling nodes that may be undesirable, especially when surrounded by other inline elements or non-whitespace text.

Another cause of likely unwanted wrapping is when an inline element is present as a child element.

Here is an example, similar to #14:

```diff
--- input.html	2019-05-12 17:17:01.000000000 -0600
+++ output.html	2019-05-12 17:17:01.000000000 -0600
@@ -1,3 +1,5 @@
 <div>
-    [<span>{score}</span>]
+    [
+    <span>{score}</span>
+    ]
 </div>
```

---

Perhaps the easiest way to rectify both issues is to hardcode a list of inline elements when checking if a node is an inline node:

https://github.com/UnwrittenFun/prettier-plugin-svelte/blob/771df545e15f422a9597e9153a1518dbbba8c079/src/print/index.ts#L467-L469

This will cause the element to be treated as part of the inline sequence `fill` groups already used in `printChildren` – this will allow the element's tags to remain in its current position and only break between words when necessary, not forcibly before the start tag and at the start and end of the element's child content.

For example:

```diff
--- input.html	2019-05-12 17:30:45.000000000 -0600
+++ output.html	2019-05-12 17:30:45.000000000 -0600
@@ -1 +1,3 @@
-<span>Lorem ipsum dolor sit amet consectetur adipisicing elit. Tenetur ad fugit cupiditate vel ullam voluptatum maxime harum nemo ipsa sequi sapiente assumenda vitae, beatae iste sed veritatis ipsam dicta deserunt.</span>
+<span>Lorem ipsum dolor sit amet consectetur adipisicing elit. Tenetur ad fugit
+    cupiditate vel ullam voluptatum maxime harum nemo ipsa sequi sapiente
+    assumenda vitae, beatae iste sed veritatis ipsam dicta deserunt.</span>
```

This (obviously) has the disadvantage of making assumptions about the the `display` property of all elements, but (in my opinion) should handle most cases where undesired wrapping currently occurs.

---

What is your opinion on this? Is this something you want to try to handle? Should we go with the easy route, or should we try to replicate the (very complicated) logic used for vanilla HTML by Prettier core?

### LINKED ISSUE #24: HTML whitespace is not maintained.
Wrapping a word just before punctuation, but not wrapping the punctuation, results in bad formatting.

```html
<p>This is a <strong>bold stance</strong>.</p>
```
becomes

```html
<p>
  This is a
  <strong>bold stance</strong>
  .
</p>
```

Which is rendered as:

This is a **bold stance** .

Notice the space before the period.

### LINKED ISSUE #28: Configure whitespace-sensitive tags
Related to #24, whitespace-sensitive tags should not be modified on format. For example, this

```html
    <pre>foo
bar baz</pre>
```

becomes this

```html
    <pre>foo bar baz</pre>
```

I would also like to be able to configure custom tags to follow the same rules as the `<pre>` tag. For example, if I configure the `CodeSnippet` tag to be whitespace-sensitive, this should not change after formatting:

```html
    <CodeSnippet lang="js">const foo = () => {
  return 42
}</CodeSnippet>
```

### PR DESCRIPTION
**Pull request by jarrodldavis**
I'm opening this pull request early to get feedback on the tests I added before finishing the fixes I had in mind.

- [x] Add additional tests for formatting/printing inline tags, including inside other tags and alongside text node siblings
- [x] Add test for printing (and not reformatting) preformatted tags like `<pre>`
- [ ] Any additional tests as necessary
- [x] Prevent wrapping before and after inline tags (fixes #20, fixes #24)
- [x] Allow customization of what tags are considered inline
- [x] Prevent formatting inside pre-formatted tags like `<pre>` (fixes #28)
- [x] Allow customization of what tags are considered preformatted

The following seem like they will be hard to do until a `string` option type is provided by Prettier (see https://github.com/prettier/prettier/issues/6151):
- [ ] Allow additive customization of tags that are considered preformatted – mainly to allow for configuring components without overriding configuration of vanilla HTML tags
- [ ] Allow additive customization of tags that are considered inline – mainly to allow for configuring components without overriding configuration of vanilla HTML tags

### CODE REVIEW

File: test/printer/samples/inline-element-with-element-mustache.html [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<a>
+    <MyIcon />
```
Comment (jarrodldavis) [inline test/printer/samples/inline-element-with-element-mustache.html]: How should we handle block elements inside inline elements? Should they wrap as normal?

Comment (UnwrittenFun) [reply test/printer/samples/inline-element-with-element-mustache.html]: yeah I think wrap as normal

**Comment by UnwrittenFun** [comment]
Looks good so far 👍

File: test/printer/samples/inline-element-with-element-mustache.html (hunk 1 above)
Comment (jarrodldavis) [reply test/printer/samples/inline-element-with-element-mustache.html]: Hmm, so it looks like Prettier handles this differently (with `htmlWhitespaceSensitivity` set to `strict` or `css`): even with child block tags, it doesn't break immediately after the opening tag nor immediately before the closing tag.

[Here are some examples]

Coincidentally, that behavior is the easiest way to implement the fixes for inline elements in this plugin. [Here's the updated behavior]

[Here are some examples]: https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAJgSwG4D4AWmqA9FnhjgXADbUQlm4A6ULqAzgA4CGUuAnvWJdeuDjz4xCUAOYkRfFmwUEijfDToNKheROaso40RTwbaQxqdXbyw-SAA0ICJxiZo7ZKG4AnXxAA7gAKfgheKNzYEJjoTiAARr7cYADWcDAAyjxgmLLIML4ArnDO+DAAttQA6oTwImBwmeGY7tit-Mjg7F7OeexwvjDByTIV3MgAZtzUA84AVuwAHgBCyWkZmdwVcAAyeXBTM3Mgi0uZeTLUcACKRRDwR7OlIDy+A75d5VXxnL55MGqsSkyAAHAAGZx-CADarJThdP5wD7YQ7OACO93gI1cERA3HYAFooHA4OgyfFfHBMZgqSNuGMJkhps9nAMKpgCsUXuxLtc7g9Dszji8YNwEkD0CCkAAmZyFbiYaiXADCEAq4y6UGgaJARQGABVxREWQMAL5moA

[Here's the updated behavior]: https://github.com/UnwrittenFun/prettier-plugin-svelte/pull/41/commits/a599244c1563a59baaac7af5a52b34a0f77c83c4

**Comment by lukeed** [comment]
This is a great PR. What else needs to be done?

**Comment by jarrodldavis** [comment]
Hey sorry, it's been a while since I've looked at this. From my checklist, it seems I was stuck on making the configuration of inline/preformatted tags as flexible as possible since Prettier didn't have a proper `string` type for options, but it appears that has since been fixed: https://github.com/prettier/prettier/pull/6219

Also, maybe more tests? 😬

**Comment by awmottaz** [comment]
@jarrodldavis can I help get this PR across the finish line?

**Comment by jarrodldavis** [comment]
@awmottaz Feel free to fork/copy over my branch and make your own PR based on this one. I don't have the time right now to finish this, sorry.

**Pull request closed without merging**